### PR TITLE
[Rust] [Experiment] Trigonometry kernels

### PR DESCRIPTION
**Pull request by nevi-me**
This is on top of #9297

I was curious if (ab)using the `compute::unary` kernel would perform better on slightly complex functions.

I implemented the Haversine function, which calculates the distance between two geographic coordinates.
I then benchmarked an implementation that I tried to simplify and optimise with unary kernels, vs one that I'd have to write if I couldn't use the unary kernels for things like:
- arithmetics with scalars
- functions that would otherwise require generating intermediate arrays (e.g. `sin(x) * cos(x)` would be `multiply(sin(x), cos(x))`)

The function that uses unary kernels for the above, is slightly faster.

I ran this on an M1 CPU, with the below options

```sh
cargo bench --bench trigonometry_kernels
cargo bench --bench trigonometry_kernels --features simd
RUSTFLAGS="-C target-cpu=native" cargo bench --bench trigonometry_kernels
RUSTFLAGS="-C target-cpu=native" cargo bench --bench trigonometry_kernels --features simd
```

```rust
haversine_no_unary 512  time:   [14.074 us 14.140 us 14.216 us]

haversine_unary 512     time:   [11.191 us 11.308 us 11.436 us]
haversine_no_unary_nulls 512                                                                             
                        time:   [15.902 us 15.985 us 16.083 us]

haversine_unary_nulls 512                                                                             
                        time:   [12.486 us 12.552 us 12.625 us]
```

The biggest benefit is from setting the `RUSTFLAGS`, the non-null benches go 3-10% faster.

### CODE REVIEW

File: rust/arrow/src/compute/kernels/trigonometry.rs [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+{
+    // Check array lengths, must all equal
+    let len = lat_a.len();
+    if lat_b.len() != len || lng_a.len() != len || lng_b.len() != len {
+        return Err(ArrowError::ComputeError(
+            "Cannot perform math operation on arrays of different length".to_string(),
+        ));
+    }
+    // These casts normally get optimized to f64 as f64
+    let one = num::cast::<_, T::Native>(1.0).unwrap();
```
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/trigonometry.rs:110]: @jorgecarleitao I had to convert the floats to `T::Native` in order to use them in the unary kernel as part of `Fn(_)` expressions.

Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/trigonometry.rs:110]: I did not find any easy way of doing this, also. It feels like we would need a generic of a generic. It looks good to me, though.

**Comment by jorgecarleitao** [comment]
IMO looked good to me (before the rebase), however, a more through review would need a rebase xD

**Comment by nevi-me** [comment]
I've rebased.

Ideas for the PR before I work more on it:

- Are the trig kernels fine (is the direction good)?
- Should I remove Haversine? I was using it to evaluate the performance benefit of using `unary`

File: rust/arrow/src/compute/kernels/trigonometry.rs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        )?,
+    )?;
+    Ok(unary::<_, _, _>(
+        &atan2(&sqrt(&a), &sqrt(&unary::<_, _, _>(&a, |x| one - x)))?,
+        |x| x * two_radius,
+    ))
+}
+
+// only added here to test that it's accurate
+pub fn haversine_no_unary<T>(
```
Comment (jorgecarleitao) [inline rust/arrow/src/compute/kernels/trigonometry.rs:141]: shouldn't we move this to the tests, if it is only used for testing?

File: rust/arrow/benches/trigonometry_kernels.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+use arrow::error::{ArrowError, Result};
+use arrow::util::test_util::seedable_rng;
+use arrow::{
+    array::*,
+    compute::*,
+    datatypes::{ArrowNumericType, ArrowPrimitiveType},
+};
+
+/// This computes the Haversine formula without using unary kernels
+fn haversine_no_unary<T>(
```
Comment (jorgecarleitao) [inline rust/arrow/benches/trigonometry_kernels.rs:37]: I would try to get the result from the exercise, drop this code once the best option was found, and document which was was better in implementation notes on the decided approach.

**Review by jorgecarleitao — APPROVED** [review]
I am not sure if this is ready to review, as it is marked as draft, but I reviewed it anyways.

LGTM, thanks a lot for taking this! :)

I left some ideas / suggestions but I am fine as is also.

**Comment by nevi-me** [comment]
Thanks @jorgecarleitao, I'm currently constrained with time during the week. I'll finish this up over the weekend

File: rust/arrow/src/compute/kernels/trigonometry.rs (hunk 2 above)
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/trigonometry.rs:141]: I've removed it, as it's also in the benchmark

File: rust/arrow/src/compute/kernels/math.rs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+use crate::datatypes::ArrowNumericType;
+use crate::{array::*, float_unary};
+
+use super::arity::unary;
+
+float_unary!(sqrt);
+
+pub fn powf<T>(array: &PrimitiveArray<T>, n: T::Native) -> PrimitiveArray<T>
+where
+    T: ArrowNumericType,
```
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/math.rs:31]: This constraint might be incorrect. A `i32` shouldn't have `powf({float})`

**Comment by alamb** [comment]
@nevi-me I am closing this PR for the time being to clean up the Rust/Arrow PR backlog.  Please let me know if this is a mistake

**Pull request closed without merging**